Updating AQM to CMAQv5.4

drnimbusrain opened the pull request:
<!--Instructions: All subsequent sections of text should be filled in as appropriate.-->
## PR Checklist
- [ ] This PR has been tested on an RDHPCS machine and/or WCOSS2. Please select below:
  - [X ] RDHPCS.
  - [ X] WCOSS2. (Issues with output total PM2.5 fractions and calculation, equal to zero only on WCOSS2) Issue has been made on UWM by Jianping:  https://github.com/ufs-community/ufs-weather-model/issues/2378

- [X ] This PR has been tested with the ufs-srweather-app workflow [aqm_dev] and [develop] branch.

- [X ] New or updated input data is required by this PR. <!-- If checked, please work with the code managers to update input data sets on all platforms.-->
To test this PR with CMAQv5.4 in AQMv7, you will need the following:

- Updated [aqm.rc template](https://github.com/NOAA-EMC/AQM/compare/develop...noaa-oar-arl:AQM:feature/cmaq54?expand=1) file that has updated mechanism (CB6r5AE7) and some other updates.  
- Updated [diag_table_aqm.FV3_GFS_v6](https://github.com/noaa-oar-arl/ufs-srweather-app/blob/feature/cmaq54/parm/diag_table_aqm.FV3_GFS_v16) and [field_table_aqm.FV3_GFS_v16](https://github.com/noaa-oar-arl/ufs-srweather-app/blob/feature/cmaq54/parm/diag_table_aqm.FV3_GFS_v16) to accommodate CMAQv5.4 changes.

- [ X] Baselines are expected to change.

## Description
This PR provides all necessary updates to AQM develop branch for ARL's work to update CMAQv5.4 submodule (from current CMAQv5.2.1) and all necessary AQM infrastructure to accommodate the CMAQv5.4 changes compared.  Since our CMAQv5.4 changes were built upon our fork of the EMC develop branch, then it also contains some initial developments of our in-canopy photolysis infrastructure and code developments.  The canopy option is turned off by default, and at this time, we recommend that the canopy option remains turned OFF (in aqm.rc) until further refined developments of the canopy codes in AQM (and later in UFS-SRW-App, UWM, FV3/CCPP, etc. via separate corresponding PRs to their authoritative repos to enable full functionality). 

## Issue(s) addressed
#100 

## Dependencies
 Updates to UFS-SRW-App at non-default branches (e.g., production/AQM.v7) are needed to accommodate the CMAQv5.4 updates in AQM (see new or updated data section above). The question remains as to where and how to make these PRs.  Our changes for CMAQv5.4 were built upon the previous [aqm_dev] branch of UFS-SRW-App, and do not correspond to the production/AQM.v7 branch.  **Please advise and we can update and make the correct PR to the correct UFS-SRW-App branch to accommodate testing the CMAQv5.4 updates to AQM.** 

@lwcugb @bbakernoaa @ytangnoaa

Commit:
Initial commit of rrfs_cmaq_canopy branch.

Commit:
Initial copy of canopy photolysis routines.

Commit:
Initial modifications to canopy photolysis CMAQ5.2.1 codes.

Commit:
Updated CMAQ makefiles for modified canopy codes.

Commit:
Added new canopy variables to AQM shared components.

Commit:
Added placeholder new canopy variables to aqm cap for fv3.

Commit:
Added conditional CANOPY_SHADE environment variable/logical.

Commit:
Updated conditional canopy_yn environment and logicals.

Commit:
Fixed bugs.

Commit:
Fixed more bugs.

Commit:
Removed "RETURN" bug and added diagnostic prints.

Commit:
Removed RETURN bug.

Commit:
Added debug statements

Commit:
Removed debug prints and added canopy variables in DESC3.

Commit:
Added some debug prints.

Commit:
Updated debug statemetns

Commit:
Fixed bug in declaration.

Commit:
Checking CANOPY_SHADE condition..

Commit:
Updated Canopy debugs

Commit:
Fixed CANOPY_SHADE logic bug and added debug prints.

Commit:
Removed extraneous debug prints.

Commit:
Removed debug prints.

Commit:
Fixed allocation/save bug.

drnimbusrain committed:
Merge pull request #1 from noaa-oar-arl/feature/rrfs_cmaq_canopy

Feature/rrfs cmaq canopy

drnimbusrain committed:
Update Makefile.in

Fixed Makefile.in typo.

drnimbusrain committed:
Update Makefile.in

drnimbusrain committed:
Added new canopy file to read for AQM.

drnimbusrain committed:
Updated aqm.rc example file.

drnimbusrain committed:
Updated example aqm.rc for canopy settings and file.

drnimbusrain committed:
Initial changes for reading canopy data in AQM.

drnimbusrain committed:
Updated aqm_emis_read and aqm.rc for canopy variables.

drnimbusrain committed:
Updated bug to get aqm_get_config.

drnimbusrain committed:
Merge branch 'NOAA-EMC:develop' into feature/aqm_canopy

drnimbusrain committed:
Updated for local in-canopy modified codes.

drnimbusrain committed:
Moved ASX_DATA_MOD to compile above Phot.F

drnimbusrain committed:
Updated to remove default ASX_DATA_MOD and phot.F

bbakernoaa committed:
updates

bbakernoaa committed:
updates

bbakernoaa committed:
Revert "updates"

This reverts commit 1abdda056abb38083f9425d4ed38a88b37af3aea.

bbakernoaa committed:
Revert "updates"

This reverts commit 43588af77cf86ee9f24cc67437d625d0ebede984.

ytangnoaa committed:
initialize feature/pt-source

drnimbusrain committed:
Merge branch 'NOAA-EMC:develop' into feature/aqm_canopy

drnimbusrain committed:
Testing Sub-Canopy phot effects only.

rmontuoro committed:
Enable CMAQ simplistic scavenging and wet removal in resolved clouds.

rmontuoro committed:
Fix naming inconsistency for convective and nonconvective precipitation fields.

rmontuoro committed:
Properly set compiler flags in GNU build system for fixed source form files.

rmontuoro committed:
Merge branch 'develop' into feature/aqm-rescld-wdep

ytangnoaa committed:
enable point source per DE

ytangnoaa committed:
Merge pull request #4 from NOAA-EMC/develop

merge with Fengsha

ytangnoaa committed:
minor update

drnimbusrain committed:
Merge pull request #7 from NOAA-EMC/develop

Merge EMC develop to ARL develop

drnimbusrain committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/AQM into feature/aqm_canopy

drnimbusrain committed:
Merge pull request #8 from noaa-oar-arl/feature/aqm_canopy

Feature/aqm canopy merge to develop

ytangnoaa committed:
Merge pull request #5 from noaa-oar-arl/feature/pt-source

Feature/pt source

zmoon committed:
Merge pull request #9 from rmontuoro/feature/aqm-rescld-wdep

Cloud scavenging

zmoon committed:
Merge branch 'NOAA-EMC:develop' into develop

zmoon committed:
Wet deposition fix from @rmontuoro

based on the current diff of
/scratch2/NCEPDEV/naqfc/Raffaele.Montuoro/flux/dev/pr/jianping/dev/ufs-weather-model/AQM/src/shr/aqm_methods.F90

(couldn't just copy the file since we have other changes)

zmoon committed:
Revert "Wet deposition fix from @rmontuoro"

This reverts commit 71ebc0af939319f97e9369e25961c2643461bcef.

So that we can merge the official version more easily

zmoon committed:
Merge remote-tracking branch 'upstream/develop' into develop

Had to manually resolve some conflicts, mainly in the makefile stuff
in order to keep canopy, but also in aqm.rc (to keep canopy stuff)

zmoon committed:
Merge remote-tracking branch 'upstream/develop' into develop

This required some manual merge conflict resolution, mainly for
the point source stuff, as it seems like they didn't start from
Youhua's branch.

Also note that `fengsha_yn` is no longer part of the config object
as it is now detected another way via CMAQ emissions config

zmoon committed:
Fix merge bug

duplicate cases in select

zmoon committed:
Remove leftover code in PT3D_DEFN

from the initial PT impl

zmoon committed:
Remove duplicate (and non-guarded) `CLDPROC` call

seems was leftover when the new guarded one was merged in

zmoon committed:
style

reduce diff wrt. upstream

zmoon committed:
Remove duplicate get-config in `aqm_methods`

zmoon committed:
Remove 'AREA' case in `interpx`

was added by Youhua but is not present in upstream

zmoon committed:
Remove fenghsa `em%dens_flag` block

upstream doesn't have it

drnimbusrain committed:
Uncommenting diagnostic prints phot.F

Quick check of canopy inputs and photolysis attenuation factors.

drnimbusrain committed:
Update phot.F

drnimbusrain committed:
Update phot.F

Fixed bug on integrating though ZCAN=ZFL, and rolled back to FCH > 0.5 m canopy condition to be consistent with initial implementation.

drnimbusrain committed:
Update phot.F

Fixed bug on integrating though ZCAN=ZFL, and rolled back to FCH > 0.5 m canopy condition to be consistent with initial implementation.

BrianCurtis-NOAA committed:
Revert fixes to get debug mode working with UFSWM debug run (#57)

drnimbusrain committed:
Restricted to FCH and LAI only to match CCPP physics.

drnimbusrain committed:
Fixed bug in IF statement for canopy conditions.

drnimbusrain committed:
Merge branch 'develop' into feature/aqm_canopy

drnimbusrain committed:
Merge pull request #14 from noaa-oar-arl/feature/aqm_canopy

Feature/aqm canopy

zmoon committed:
Merge branch 'prod' into develop

zmoon committed:
Merge remote-tracking branch 'upstream/develop' into develop

bbakernoaa committed:
Update CMAQ

ytangnoaa committed:
update CMAQ respository

lwcugb committed:
update aqm_files.cmake with corresponding code changes in the local /model/src folder

lwcugb committed:
Merge pull request #15 from lwcugb/feature/cmaq54_li

update aqm_files.cmake with corresponding code changes in the local /…

btang1 committed:
Add files via upload

Beiming new DUST_EMIS for ufs-cmaq54

btang1 committed:
Add files via upload

btang1 committed:
Add files via upload

btang1 committed:
Update centralized_io_module.F

btang1 committed:
Add files via upload

un-commented all MAPS lines, and changed CMakeList.txt

btang1 committed:
Add files via upload

commented MPAS in CMakeLists.txt

ytangnoaa committed:
Merge pull request #16 from btang1/feature/cmaq54_beiming

Add files via upload

btang1 committed:
Add files via upload

btang1 committed:
Add files via upload

btang1 committed:
Add files via upload

btang1 committed:
Add files via upload

btang1 committed:
Update ASX_DATA_MOD.F

btang1 committed:
Update o3totcol.f

btang1 committed:
Update vdiffacmx.F

comment a 'end do', after that, code compiled successfully

btang1 committed:
Update aqm_files.cmake

turn on according changed codes in local src/ folder

ytangnoaa committed:
Merge pull request #17 from btang1/feature/cmaq54_beiming

Add files via upload

lwcugb committed:
add localized files

lwcugb committed:
Add other modified files

lwcugb committed:
Add other modified files

lwcugb committed:
Update .gitmodules

point to CMAQ 5.4+

lwcugb committed:
Update aqm_methods.F90

Forget to turn on AOD

drnimbusrain committed:
Merge pull request #20 from lwcugb/feature/cmaq54_li

Feature/cmaq54 li

lwcugb committed:
Update RUNTIME_VARS.F

Need to comment out two lines for writing to OUTDEV and LOGDEV in aqm_dev branch

lwcugb committed:
Change aqm_files.cmake to use ROS3 solver

lwcugb committed:
Add rbdriver.F file to localCCTM folder. There is a typo in CMAQ's source file and therefore need to put it here

ytangnoaa committed:
Merge pull request #23 from lwcugb/feature/camq54_ros3

Feature/camq54 ros3

JianpingHuang-NOAA commented:
@BrianCurtis-NOAA  Can you merge this PR?

BrianCurtis-NOAA commented:
> @BrianCurtis-NOAA Can you merge this PR?

I think there is other work going on that will incorporate a bunch of things to get CMAQ 5.4 working. Let's get this figured out.

ytangnoaa commented:
> @BrianCurtis-NOAA Can you merge this PR?

Jianping. Please hold it. We may update it for some bugs.

lwcugb committed:
Add files via upload

Better handle of the emission unit. This will not impact the simulation results.

lwcugb committed:
Update PT3D_STKS_DEFN.F

Resolve a bug in point source emission. Only the first layer emission is included with his bug, which leads to much lower SO2 simulations.

lwcugb committed:
Delete some comments for testing purpose

ytangnoaa committed:
Merge pull request #24 from lwcugb/feature/cmaq54

Fix a bug in point source emission

JianpingHuang-NOAA commented:
@ytangnoaa@lwcugb  Do you have any preliminary results to share by using the latest code?

BrianCurtis-NOAA commented:
This supersedes #98 
Could you change line 23 of CMakeLists.txt to: `set(CMAKE_Fortran_FLAGS_DEBUG "-O0 -g -check all -check noarg_temp_created -fpe0 -fno-alias -ftrapuv -traceback")`

ytangnoaa committed:
Update CMakeLists.txt

ytangnoaa commented:
> This supersedes #98 Could you change line 23 of CMakeLists.txt to: `set(CMAKE_Fortran_FLAGS_DEBUG "-O0 -g -check all -check noarg_temp_created -fpe0 -fno-alias -ftrapuv -traceback")`

Updated the compiler flag.

BrianCurtis-NOAA commented:
If you could test a build it would save me some time later today, but there might be some tweaks to some of the code to conform to those options.

JianpingHuang-NOAA commented:
@BrianCurtis-NOAA @bbakernoaa  @drnimbusrain  Is this PR ready to be merged ?

bbakernoaa commented:
@JianpingHuang-NOAA @BrianCurtis-NOAA  it looks like we still need to add a test for this.  Could you help with this @JianpingHuang-NOAA?

BrianCurtis-NOAA commented:
> @JianpingHuang-NOAA @BrianCurtis-NOAA it looks like we still need to add a test for this. Could you help with this @JianpingHuang-NOAA?

We need a test, yes. It also needs to be one that can run quickly. C96 preferrably, but 192 is acceptable if it makes things easier. The test needs to run well within a 30 minute wall clock. It needs to test as many features as it can reasonably. If there are other features that need testing that can't be added to one test, then two can be acceptable if necessary.

BrianCurtis-NOAA commented:
> > @JianpingHuang-NOAA @BrianCurtis-NOAA it looks like we still need to add a test for this. Could you help with this @JianpingHuang-NOAA?
> 
> We need a test, yes. It also needs to be one that can run quickly. C96 preferrably, but 192 is acceptable if it makes things easier. The test needs to run well within a 30 minute wall clock. It needs to test as many features as it can reasonably. If there are other features that need testing that can't be added to one test, then two can be acceptable if necessary.

I hope that eventually the workflow would be able to generate a run directory suited for regression testing so that future adjustments to the UFSWM regression tests are much easier.

JianpingHuang-NOAA commented:
@bbakernoaa, https://github.com/noaa-oar-arl/AQM/tree/feature/cmaq54 is 3-month old.  Where is the updated package?

bbakernoaa commented:
this is the latest that we have.  We don't have anything newer at this juncture

JianpingHuang-NOAA commented:
I heard that ARL has a fix to the soil type mix-up. Is that not for AQM?

bbakernoaa commented:
This is something that needs to be changed not just on the AQM side but also the FV3ATM side as well.  We just finished some preliminary testing and such today.  We will need to do PRs back to the UFS weather model for both changes (ie AQM and FV3ATM repositories)

bbakernoaa commented:
note that when we update to the GFS v17 physics we should also pass the fractional land use over and use this instead of the dominant value.  Currently we are basically hard coding the fractional land use to be 1 for each dominate land use type in the grid cell and all other values are 0.

drnimbusrain commented:
I agree with Barry.  We should handle this AQM/CMAQ54 update here, and
leave other updates including physics, bug fixes, etc in other separate PRs
as suitable.

On Fri, Jan 3, 2025, 2:04 PM Barry Baker ***@***.***> wrote:

> note that when we update to the GFS v17 physics we should also pass the
> fractional land use over and use this instead of the dominant value.
> Currently we are basically hard coding the fractional land use to be 1 for
> each dominate land use type in the grid cell and all other values are 0.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/AQM/pull/101#issuecomment-2569687291>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGLFYNR5BF7MX6PJQOLMMDL2I3NMRAVCNFSM6AAAAABLA6CBLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNRZGY4DOMRZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

drnimbusrain commented:
@BrianCurtis-NOAA How's it going moving this CMAQ54 PR forward, and is there a status update on a regression test?  Do you need help from us to identify inputs, run directory, etc. to help get this setup and moved forward?  
@JianpingHuang-NOAA

BrianCurtis-NOAA commented:
@drnimbusrain We will need C96 support for AQM to get a proper regression test (C192 is acceptable if C96 complicates things)

I don't need a workflow to get this merged, but a run directory with a working UFSWM will suffice so I can update the RT system to use it. As long as we have that, we can get this part merged in. Is this something that you might be able to put together?

bbakernoaa commented:
@BrianCurtis-NOAA could we not leverage: https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/tests/regional_atmaq

We would just need to update the additional CMAQ related inputs and such for the updates no?

BrianCurtis-NOAA commented:
> @BrianCurtis-NOAA could we not leverage: https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/tests/regional_atmaq
> 
> We would just need to update the additional CMAQ related inputs and such for the updates no?

Yes, that can be used as a template.

BrianCurtis-NOAA commented:
I'm happy to put the work in taking a run directory and turning it into a proper RT if that will save you from too much work. The work on AQM-Workflow will eventually know how to generate such a run directory, but that will be a while.

drnimbusrain commented:
Thanks Brian.  That link to regional_atmaq RT seems pretty old, yes?  So,
to confirm you need a C96 run directory with AQM CMAQ54 updates that you
will use as a template for RT? We mainly have been running C793.

On Fri, Jan 17, 2025, 4:01 PM Brian Curtis ***@***.***> wrote:

> I'm happy to put the work in taking a run directory and turning it into a
> proper RT if that will save you from too much work. The work on
> AQM-Workflow will eventually know how to generate such a run directory, but
> that will be a while.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/AQM/pull/101#issuecomment-2599201849>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGLFYNSO2MYVIQ6VIYAA3CL2LFVSPAVCNFSM6AAAAABLA6CBLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGIYDCOBUHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bbakernoaa commented:
@drnimbusrain I think all we need to do is provide a run directory. The truth is most of the input files as far as emissions and such can be used for either 5.2 or 5.4. We just need to make sure we have an example to show the differences in the ancillary CMAQ files needed. Other than that the input files may need updated on the ATM side to accommodate the updates of the newer ATM hash. 

I can take a quick stab at it on Tuesday morning. At least to get something running. Point me to a run directory that you have for whatever resolution. Then @BrianCurtis-NOAA you could then integrate the updated files that RT needs into run RT workflow.

drnimbusrain commented:
Great news, thanks Barry.  I'll send a run directory we have soon.

On Fri, Jan 17, 2025, 9:43 PM Barry Baker ***@***.***> wrote:

> @drnimbusrain <https://github.com/drnimbusrain> I think all we need to do
> is provide a run directory. The truth is most of the input files as far as
> emissions and such can be used for either 5.2 or 5.4. We just need to make
> sure we have an example to show the differences in the ancillary CMAQ files
> needed. Other than that the input files may need updated on the ATM side to
> accommodate the updates of the newer ATM hash.
>
> I can take a quick stab at it on Tuesday morning. At least to get
> something running. Point me to a run directory that you have for whatever
> resolution. Then @BrianCurtis-NOAA <https://github.com/BrianCurtis-NOAA>
> you could then integrate the updated files that RT needs into run RT
> workflow.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/AQM/pull/101#issuecomment-2599483757>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGLFYNSMACKRC4MCPTKEC532LG5TJAVCNFSM6AAAAABLA6CBLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGQ4DGNZVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bbakernoaa commented:
@BrianCurtis-NOAA @drnimbusrain please take a look here: `/scratch1/NCEPDEV/stmp2/Barry.Baker/FV3_RT/rt_994030/regional_atmaq_debug_intel`



I was able to get the RT for the regional_atmaq_debug going on hera with the updated feature/cmaq54 branch of the AQM.  Note that this was off the latest develop branch of the UFSWM 

### Files Changed 

```
AE_cb6r3_ae6_aq.nml -> AE_cb6r5_ae7_aq.nml
CSQY_DATA_cb6r3_ae6_aq -> CSQY_DATA_cb6r5_ae7_aq
GC_cb6r3_ae6_aq.nml -> GC_cb6r5_ae7_aq.nml
NR_cb6r3_ae6_aq.nml -> NR_cb6r5_ae7_aq.nml
PHOT_OPTICS.dat
Species_Table_TR_0.nml
field_table
diag_table
aqm.rc 
CMAQ_Control_DESID.nml
CMAQ_Control_DESID_cb6r5_ae7_aq.nml
CMAQ_Control_Misc.nml
```

BrianCurtis-NOAA commented:
are you using CCPP_SUITES=FV3_GFS_v16 ?

bbakernoaa commented:
the RT was using `FV3_GFS_v15p2`

I'll try to quickly update it to `FV3_GFS_v16`

bbakernoaa commented:
@BrianCurtis-NOAA I updated it to use the `FV3_GFS_v16` physics

JianpingHuang-NOAA commented:
@BrianCurtis-NOAA , what is the status of the PR merging ?

BrianCurtis-NOAA commented:
@JianpingHuang-NOAA It will be a little bit. I'll get to looking at Barry's run dir soon.

JianpingHuang-NOAA commented:
@BrianCurtis-NOAA  Hi Brian, any update on this?

JianpingHuang-NOAA commented:
@BrianCurtis-NOAA  Hi Brian, we need to get this PR merged as soon as possible. Thanks !

BrianCurtis-NOAA commented:
@bbakernoaa I don't see the run directory you have listed above any more. Did you move it or have it somewhere else?

BrianCurtis-NOAA commented:
> @bbakernoaa I don't see the run directory you have listed above any more. Did you move it or have it somewhere else?

typo, nevermind

BrianCurtis-NOAA commented:
There have been some issues running the model with a debug build, with the most recent issue:
```
WARNING: map_precursor: NO particulate mercury in model.       
forrtl: severe (408): fort: (3): Subscript #1 of the array WFC has value 0 which is less than the lower bound of 1

Image              PC                Routine            Line        Source             
fv3.exe            000000000BAA303F  Unknown               Unknown  Unknown
fv3.exe            000000000218451F  asx_data_mod_mp_i         840  ASX_DATA_MOD.F
fv3.exe            000000000210AD01  vdiff_                    236  vdiffproc.F
fv3.exe            0000000001CA856E  cmaq_mod_mp_cmaq_         184  cmaq_mod.F90
fv3.exe            0000000001CA5E8D  cmaq_model_mod_mp         139  cmaq_model_mod.F90
fv3.exe            0000000001C1FC97  aqm_comp_mod_mp_a         219  aqm_comp_mod.F90
fv3.exe            0000000001C1DDE8  aqm_mp_modeladvan         594  aqm_cap.F90
```

Here is a diff of the changes that have been made so far
```diff
diff --git a/src/model/src/centralized_io_module.F b/src/model/src/centralized_io_module.F
index c0605e6..e8d5b3a 100644
--- a/src/model/src/centralized_io_module.F
+++ b/src/model/src/centralized_io_module.F
@@ -1670,7 +1670,7 @@
              if (NLDNSTRIKE) then
                  CALL NEXTIME ( loc_jdate(f_ltng), loc_jtime(f_ltng), file_tstep(f_ltng) )
              end if
-             CALL NEXTIME ( loc_jdate(f_bcon), loc_jtime(f_bcon), file_tstep(f_bcon) )
+!             CALL NEXTIME ( loc_jdate(f_bcon), loc_jtime(f_bcon), file_tstep(f_bcon) )
 #endif
 
              do i = 1, N_FILE_GR
diff --git a/src/shr/aqm_methods.F90 b/src/shr/aqm_methods.F90
index 0098d07..5754a61 100644
--- a/src/shr/aqm_methods.F90
+++ b/src/shr/aqm_methods.F90
@@ -682,18 +682,20 @@ logical function interpx( fname, vname, pname, &
     if (aqm_rc_check(localrc, msg="Failure to retrieve grid coordinates", &
       file=__FILE__, line=__LINE__)) return
 
-    if (vname(1:7) == 'LUFRAC_') then
-      lu_index = 0
-      read(vname(8:9), *, iostat=localrc) lu_index
-      if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
-        file=__FILE__, line=__LINE__)) return
-      k = 0
-      do r = row0, row1
-        do c = col0, col1
-          k = k + 1
-          if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
+    if (len(vname) > 7 ) then
+      if (vname(1:7) == 'LUFRAC_') then
+        lu_index = 0
+        read(vname(8:9), *, iostat=localrc) lu_index
+        if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
+          file=__FILE__, line=__LINE__)) return
+        k = 0
+        do r = row0, row1
+          do c = col0, col1
+            k = k + 1
+            if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
+          end do
         end do
-      end do
+      end if
     else
       select case (trim(vname))
         case ('HT')
@@ -1147,18 +1149,20 @@ LOGICAL FUNCTION  XTRACT3 ( FNAME, VNAME,                           &
 !      end do
 !    end if
 
-    if (vname(1:7) == 'LUFRAC_') then
-      lu_index = 0
-      read(vname(8:9), *, iostat=localrc) lu_index
-      if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
-        file=__FILE__, line=__LINE__)) return
-      k = 0
-      do r = row0, row1
-        do c = col0, col1
-          k = k + 1
-          if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
+    if (len(vname) > 7) then
+      if (vname(1:7) == 'LUFRAC_') then
+        lu_index = 0
+        read(vname(8:9), *, iostat=localrc) lu_index
+        if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
+          file=__FILE__, line=__LINE__)) return
+        k = 0
+        do r = row0, row1
+          do c = col0, col1
+            k = k + 1
+            if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
+          end do
         end do
-      end do
+      end if
     else
       select case (trim(vname))
         case ('HT')
```

ytangnoaa commented:
> There have been some issues running the model with a debug build, with the most recent issue:
> 
> ```
> WARNING: map_precursor: NO particulate mercury in model.       
> forrtl: severe (408): fort: (3): Subscript #1 of the array WFC has value 0 which is less than the lower bound of 1
> 
> Image              PC                Routine            Line        Source             
> fv3.exe            000000000BAA303F  Unknown               Unknown  Unknown
> fv3.exe            000000000218451F  asx_data_mod_mp_i         840  ASX_DATA_MOD.F
> fv3.exe            000000000210AD01  vdiff_                    236  vdiffproc.F
> fv3.exe            0000000001CA856E  cmaq_mod_mp_cmaq_         184  cmaq_mod.F90
> fv3.exe            0000000001CA5E8D  cmaq_model_mod_mp         139  cmaq_model_mod.F90
> fv3.exe            0000000001C1FC97  aqm_comp_mod_mp_a         219  aqm_comp_mod.F90
> fv3.exe            0000000001C1DDE8  aqm_mp_modeladvan         594  aqm_cap.F90
> ```
> 
> Here is a diff of the changes that have been made so far
> 
> ```diff
> diff --git a/src/model/src/centralized_io_module.F b/src/model/src/centralized_io_module.F
> index c0605e6..e8d5b3a 100644
> --- a/src/model/src/centralized_io_module.F
> +++ b/src/model/src/centralized_io_module.F
> @@ -1670,7 +1670,7 @@
>               if (NLDNSTRIKE) then
>                   CALL NEXTIME ( loc_jdate(f_ltng), loc_jtime(f_ltng), file_tstep(f_ltng) )
>               end if
> -             CALL NEXTIME ( loc_jdate(f_bcon), loc_jtime(f_bcon), file_tstep(f_bcon) )
> +!             CALL NEXTIME ( loc_jdate(f_bcon), loc_jtime(f_bcon), file_tstep(f_bcon) )
>  #endif
>  
>               do i = 1, N_FILE_GR
> diff --git a/src/shr/aqm_methods.F90 b/src/shr/aqm_methods.F90
> index 0098d07..5754a61 100644
> --- a/src/shr/aqm_methods.F90
> +++ b/src/shr/aqm_methods.F90
> @@ -682,18 +682,20 @@ logical function interpx( fname, vname, pname, &
>      if (aqm_rc_check(localrc, msg="Failure to retrieve grid coordinates", &
>        file=__FILE__, line=__LINE__)) return
>  
> -    if (vname(1:7) == 'LUFRAC_') then
> -      lu_index = 0
> -      read(vname(8:9), *, iostat=localrc) lu_index
> -      if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
> -        file=__FILE__, line=__LINE__)) return
> -      k = 0
> -      do r = row0, row1
> -        do c = col0, col1
> -          k = k + 1
> -          if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
> +    if (len(vname) > 7 ) then
> +      if (vname(1:7) == 'LUFRAC_') then
> +        lu_index = 0
> +        read(vname(8:9), *, iostat=localrc) lu_index
> +        if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
> +          file=__FILE__, line=__LINE__)) return
> +        k = 0
> +        do r = row0, row1
> +          do c = col0, col1
> +            k = k + 1
> +            if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
> +          end do
>          end do
> -      end do
> +      end if
>      else
>        select case (trim(vname))
>          case ('HT')
> @@ -1147,18 +1149,20 @@ LOGICAL FUNCTION  XTRACT3 ( FNAME, VNAME,                           &
>  !      end do
>  !    end if
>  
> -    if (vname(1:7) == 'LUFRAC_') then
> -      lu_index = 0
> -      read(vname(8:9), *, iostat=localrc) lu_index
> -      if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
> -        file=__FILE__, line=__LINE__)) return
> -      k = 0
> -      do r = row0, row1
> -        do c = col0, col1
> -          k = k + 1
> -          if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
> +    if (len(vname) > 7) then
> +      if (vname(1:7) == 'LUFRAC_') then
> +        lu_index = 0
> +        read(vname(8:9), *, iostat=localrc) lu_index
> +        if (aqm_rc_test(localrc /= 0, msg="Failure to identify LU_INDEX", &
> +          file=__FILE__, line=__LINE__)) return
> +        k = 0
> +        do r = row0, row1
> +          do c = col0, col1
> +            k = k + 1
> +            if (int(stateIn % stype(c,r)) == lu_index) buffer(k) = 1.0
> +          end do
>          end do
> -      end do
> +      end if
>      else
>        select case (trim(vname))
>          case ('HT')
> ```

The array of soil type is supposed to > 0, 
https://github.com/noaa-oar-arl/AQM/blob/feature/cmaq54/src/shr/aqm_methods.F90#L799
https://github.com/noaa-oar-arl/AQM/blob/feature/cmaq54/src/shr/aqm_methods.F90#L1264

though something could be wrong. To diagnose the issue, one method is adding some checking in
src/shr/aqm_methods.F90, like

 case ("SLTYP")
          p2d => stateIn % stype
          do r = row0, row1
            do c = col0, col1             
              if ( stateIn % stype (c,r) < 1 )  then
                if(aqm_rc_check(-1,msg=' stype wrong ')) return
              end if
            end do
          end do

 If the stype does show < 1, additional diagnose can be conducted  in

https://github.com/noaa-oar-arl/AQM/blob/feature/cmaq54/src/aqm_comp_mod.F90#L601

case ("soil_type")
            call ESMF_FieldGet(field, localDe=localDe, farrayPtr=stateIn % stype, rc=rc)
            if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, &
              line=__LINE__, &
              file=__FILE__)) &
              return  ! bail
          if (any(statein%stype<1)) then
            call ESMF_LogSetError(ESMF_RC_INTNRL_BAD, msg="Error stype < 1", &
           line=__LINE__, file=__FILE__, rcToReturn=rc)
          return
        endif

BrianCurtis-NOAA commented:
Latest error:
```
forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable STREAMFAMILYNAME when it is not allocated

Image              PC                Routine            Line        Source
fv3.exe            000000000BAB10DF  Unknown               Unknown  Unknown
fv3.exe            0000000002B757D9  desid_util_mp_des          95  desid_util.F
fv3.exe            0000000002A1041A  aero_emis_mp_desi         368  AERO_EMIS.F
```

ytangnoaa commented:
> Latest error:
> 
> ```
> forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable STREAMFAMILYNAME when it is not allocated
> 
> Image              PC                Routine            Line        Source
> fv3.exe            000000000BAB10DF  Unknown               Unknown  Unknown
> fv3.exe            0000000002B757D9  desid_util_mp_des          95  desid_util.F
> fv3.exe            0000000002A1041A  aero_emis_mp_desi         368  AERO_EMIS.F
> ```

In the file "CMAQ_Control_DESID.nml", try commenting out the namelist "Desid_SizeDist"

!&Desid_SizeDist
! Desid_Sd_nml    = 
!
! !           Stream Label   | Emiss. Mode  | Ref. Mode 
!             'ALL'          ,'FINE'        ,'FINE_REF',
!             'ALL'          ,'COARSE'      ,'COARSE_REF',
!             'WBDUST'       ,'FINE'        ,'FINE_WBDUST',
!            'WBDUST'       ,'COARSE'      ,'COARSE_WBDUST',
!             'SEASPRAY'     ,'FINE'        ,'FINE_SEASPRAY',
!             'SEASPRAY'     ,'COARSE'      ,'COARSE_SEASPRAY',
! !<Example>  'AIRCRAFT'     ,'FINE'        ,'AIR_FINE',   !To use these examples, you 
! !<Example>  'AIRCRAFT'     ,'COARSE'      ,'AIR_COARSE', ! must add entries for AIR_FINE
!                                                          ! and AIR_COARSE to the data structure
!                                                          ! em_aero_ref in AERO_DATA.!
!
!/

BrianCurtis-NOAA commented:
Merged in with #106
